tests/compat-in-docker.sh

```diff
@@ -0,0 +1,50 @@
+#!/bin/bash -eu
+
+# compat-in-docker.sh
+#
+# This file is part of mbed TLS (https://tls.mbed.org)
+#
+# Copyright (c) 2019, ARM Limited, All Rights Reserved
+#
+# Purpose
+# -------
+# This runs compat.sh in a Docker container.
+#
+# Notes for users
+# ---------------
+# If OPENSSL_CMD, GNUTLS_CLI, or GNUTLS_SERV are specified the path must
+# correspond to an executable inside the Docker container. The special
+# values "next" (OpenSSL only) and "legacy" are also allowed as shorthand
+# for the installations inside the container.
+#
+# See also:
+# - scripts/docker_env.sh for general Docker prerequisites and other information.
+# - compat.sh for notes about invocation of that script.
+
+source tests/scripts/docker_env.sh
+
+case "${OPENSSL_CMD:-default}" in
+    "legacy")  export OPENSSL_CMD="/usr/local/openssl-1.0.2g/bin/openssl";;
+    "next")    export OPENSSL_CMD="/usr/local/openssl-1.1.1a/bin/openssl";;
+    *) ;;
+esac
+
+case "${GNUTLS_CLI:-default}" in
+    "legacy")  export GNUTLS_CLI="/usr/local/gnutls-3.3.18/bin/gnutls-cli";;
+    *) ;;
+esac
+
+case "${GNUTLS_SERV:-default}" in
+    "legacy")  export GNUTLS_SERV="/usr/local/gnutls-3.3.18/bin/gnutls-serv";;
+    *) ;;
+esac
+
+run_in_docker \
+    -e M_CLI \
+    -e M_SRV \
+    -e GNUTLS_CLI \
+    -e GNUTLS_SERV \
+    -e OPENSSL_CMD \
+    -e OSSL_NO_DTLS \
+    tests/compat.sh \
```

Comment:
The existing `compat.sh` and `ssl-opt.sh` scripts don't do any `make` so I was under the impression that the `*-in-docker.sh` wrappers wouldn't either. Are you suggesting that the `*-in-docker.sh` versions go beyond this -- to be able to build the programs _in the container_ -- to have better assurance of compatibility with the container's ld and libc?

Reply:
@k-stachowiak, any thoughts? Or maybe what we really need is a `make-in-docker.sh`?

Reply:
Sorry, I forgot to address this question. I'm afraid building in the container is a must, since, as I mentioned, I ran into an actual binary compatibility issues. I think the question would rather be, how to do it.

The fact that we need to work with is, that `ssl-opt.sh` and `compat.sh` actually rely on the client and server examples having been built.

We can express it by putting make inside the docker scripts, but `make-in-docker.sh` looks like a really neat an generic solution.

@Patater This seems to require a design review. What do you think such a battery of "`...-in-docker.sh`" scripts?

Reply:
@pkolbus For `make-in-docker.sh`, what is proposed method of saving the generated binaries for re-use in a subsequent `...-in-docker.sh` script? Would the `make-in-docker.sh` script populate the host filesystem with the output? Is that counter-intuitive? Should there be a dedicated Docker volume for the output binaries?

Reply:
I've added `make-in-docker.sh` with the latest rebase. The docker argument `--volume $PWD:$PWD` (in `docker_env.sh`) causes the worktree to be shared with the container, so any build outputs in the worktree end up in the same place on the host. I've found this mechanism to work well in practice on other projects, and be intuitive (since things like logs and other failure state are easy to find and understand).

Reply:
I concur with that approach. I find it consistent with different kind of builds like `make` vs in tree `cmake`, which by default interfere with each other, but which can be worked around.

The approach suggested in this PR enables creating another script, e.g. `cmake-out-of-tree-in-docker.sh` which will do `mkdir build; cd build; cmake .. $@`, which should prevent overwriting local artifacts with the "in-docker" ones.

The bottom line is, I find this approach flexible and extensible.